cyto_dl/nn/losses/vic_reg.py

```diff
@@ -0,0 +1,42 @@
+# modified from https://github.com/facebookresearch/vicreg/blob/main/main_vicreg.py
+
+import torch
+import torch.nn.functional as F
+
+
+def off_diagonal(x):
+    n, m = x.shape
+    assert n == m
+    return x.flatten()[:-1].view(n - 1, n + 1)[:, 1:].flatten()
+
+
+class VICRegLoss(torch.nn.Module):
+    def __init__(self, num_features, sim_coeff=5, std_coeff=5, cov_coeff=1):
+        super().__init__()
+        self.num_features = num_features
+        self.sim_coeff = sim_coeff
+        self.std_coeff = std_coeff
+        self.cov_coeff = cov_coeff
+
+    def forward(self, x, y):
+        batch_size, num_features = x.shape
+
+        # view invarance loss
+        repr_loss = F.mse_loss(x, y)
+
+        x = x - x.mean(dim=0)
+        y = y - y.mean(dim=0)
+        # variance loss
+        std_x = torch.sqrt(x.var(dim=0) + 0.0001)
```

Comment:
why is this factor added here to the variance? to avoid sqrt of 0?

Reply:
yeah exactly